### ISSUE TITLE
stop token failure?

### ISSUE THREAD

**Issue by michael-conrad** [Tomiinek/Multilingual_Text_to_Speech#63]
I working on trying to get the Cherokee language working in a fork of this repository.

The amount of data is low, and I'm training with 4 other languages, however, I've run into the issue where the TTS wants to keep rambling on after the end of the pronounced text.. I'm guessing this is stop token failure?

Any suggestions?

https://github.com/CherokeeLanguage/Cherokee-TTS

**Comment by Tomiinek** [comment]
I see. There are two things:

- You may want to put more emphasis on the end prediction, see `stop_balance` or the loss weight of `stop_token`: https://github.com/Tomiinek/Multilingual_Text_to_Speech/blob/bb0855adaaa003af3df06866a31ac47ded1403f0/modules/tacotron2.py#L466-L471
- It is supposed that your training data do not contain trailing silence. If you want to add it and it has a fixed length, you need to adjust `hp.stop_frames` which tells how many frames at the end of spectrograms are considered as the end. You also need to set lower `stop_balance` in the loss formulation in this case to balance positive and negative classes.

Could you share your stop token prediction plots?

**Comment by michael-conrad** [comment]
My current tensorboard log file appears to be too large to attach? I'm trying to figure that out now. 

I'm currently adding 100ms leadin silence and 150ms leadout silence in a modified ```create_spectrograms.py``` and where any excessive lengths of silence within a recording are reduced to a max of 400ms each.

Ref: https://github.com/CherokeeLanguage/Cherokee-TTS/blob/master/prepare_spectrograms.py

This run is currently using the param file: https://github.com/CherokeeLanguage/Cherokee-TTS/blob/master/params/5d.json

Voice separation isn't working well yet either, but, I figure that can be tweaked after figuring out how to get it to stop correctly first.

**Comment by michael-conrad** [comment]
Link to log:

https://drive.google.com/drive/folders/1GKqZ7cxqdv8EJ_k-KTiK5up6iRXX_W4X?usp=sharing

**Comment by michael-conrad** [comment]
> 
>     * It is supposed that your training data do not contain trailing silence. If you want to add it and it has a fixed length, you need to adjust `hp.stop_frames` which tells how many frames at the end of spectrograms are considered as the end. You also need to set lower `stop_balance` in the loss formulation in this case to balance positive and negative classes.
> 

frames = (leadout ms * hp.sample_rate) / 1000 ?

**Comment by Tomiinek** [comment]
You are almost there, it should be `leadout_ms // hp.stft_shift_ms + 1` or something like that. The shift is 12.5 ms by default. So `hp.stop_frames = 16` should work for you.

**Comment by michael-conrad** [comment]
ok,

I'm preparing to restart the training. How many epochs before it would be worth checking on stop token working with test synthesis?

I've updated params.py to have a stop_balance parameter and updated the tacotron.py file to use this parameter instead of the hardcoded 100 that was there previously.

I've set in the 5d params file the following two new options:

```json
"stop_frames": 16,
"stop_balance": 75,
```

**Comment by Tomiinek** [comment]
Sounds good!

> How many epochs before it would be worth checking on stop token working with test synthesis? 
I am sorry, but I have no idea.

**Comment by michael-conrad** [comment]
### PREVIOUS RUN:

![image](https://user-images.githubusercontent.com/5229492/139711453-58bc24e1-21a1-4bad-be83-0a495c4e7f0b.png)

![image](https://user-images.githubusercontent.com/5229492/139711481-a911d952-586a-4b3a-8e86-8191b37b824e.png)

### CURRENT RUN:

![image](https://user-images.githubusercontent.com/5229492/139711295-06550a65-527c-40c0-b43d-6fa7c3aa862d.png)

![image](https://user-images.githubusercontent.com/5229492/139711325-19b55957-9c75-4a56-90a8-173584b12d57.png)

**Comment by Tomiinek** [comment]
This looks great! You can expect a sigmoid-like curve at the end.

**Comment by michael-conrad** [comment]
Epoch 150:

![image](https://user-images.githubusercontent.com/5229492/139855095-52a27c88-6661-44c5-9f7d-bf2de7680dc5.png)

![image](https://user-images.githubusercontent.com/5229492/139855171-b5f662ad-e5a9-45d4-9550-22d4a89cfd3f.png)

![image](https://user-images.githubusercontent.com/5229492/139855272-fc350a09-0b3e-4032-96ef-961341bf492b.png)

**Comment by Tomiinek** [comment]
The model is obviously overfitting, your train loss is 10x lower than validation. IMHO, you might want to decrease the stop token loss weight or revisit your data :shrug:

**Comment by michael-conrad** [comment]
> The model is obviously overfitting, your train loss is 10x lower than validation. IMHO, you might want to decrease the stop token loss weight or revisit your data shrug

Do you mean reduce ```stop_balance``` to say 50 or 25?

**Comment by Tomiinek** [comment]
Nono, I mean setting here: https://github.com/Tomiinek/Multilingual_Text_to_Speech/blob/bb0855adaaa003af3df06866a31ac47ded1403f0/modules/tacotron2.py#L466-L471

a weight to the stop token loss:
```
'stop_token' : weigth * F.binary_cross_entropy_with_logits(stop, target_stop, pos_weight=stop_balance) / (hp.num_mels + 2),
```
where `weigth`is you number, e.g. 0.5

**Comment by michael-conrad** [comment]
> Nono, I mean setting here:
> 
> https://github.com/Tomiinek/Multilingual_Text_to_Speech/blob/bb0855adaaa003af3df06866a31ac47ded1403f0/modules/tacotron2.py#L466-L471
> 
> a weight to the stop token loss:
> 
> ```
> 'stop_token' : weigth * F.binary_cross_entropy_with_logits(stop, target_stop, pos_weight=stop_balance) / (hp.num_mels + 2),
> ```
> 
> where `weigth`is you number, e.g. 0.5

Ah, ok, will add hyper param for that and set it to 50% for the next run. I didn't even think to compare the actual loss magnitudes between val and train. :-)

**Comment by michael-conrad** [comment]
ok, I updated my code to skip samples < 1 sec in length and adjust the stop token weight, the curve still looks like it is overfitting, but the tests I've run so far at epoch 270 have all terminated correctly.

Voice separation is still an issue though...

**Comment by michael-conrad** [comment]
Hrmm - keeping in mind that I have no clue as to what I'm doing here.

The speaker embedding dimension seems to have more of an impact on the stop loss curves than the stop loss specific parameters. A few voices don't stop properly and I am thinking that I need to get the voice separation working to be able to fix that.

I have a total of 119 speakers as follows.

chr:22
de:39
en:30
fr:22
ru:6

Thoughts?

**Comment by Tomiinek** [comment]
Hello, sorry for very late response.

This is weird. Is it possible that recordings of some of your speakers are odd? And how many recordings do you have?

**Comment by michael-conrad** [comment]
The quality of my recordings range from poor to okay.

Would training rate have any impact on this?
